Sources/PackageCollections/Storage/PackageCollectionsStorage.swift

```diff
@@ -0,0 +1,321 @@
+                try self.executeStatemnt(query) { statement -> Void in
+                    let data = try self.jsonEncoder.encode(collection)
+
+                    let bindings: [SQLite.SQLiteValue] = [
+                        .string(collection.identifier.databaseKey()),
+                        .blob(data),
+                    ]
+                    try statement.bind(bindings)
+                    try statement.step()
+                }
+                callback(.success(collection))
+            } catch {
+                callback(.failure(error))
+            }
+        }
+    }
+
+    func remove(identifier: PackageCollectionsModel.PackageCollectionIdentifier,
+                callback: @escaping (Result<Void, Error>) -> Void) {
+        self.queue.async {
+            do {
+                let query = "DELETE FROM PACKAGES_COLLECTIONS WHERE key == ?;"
+                try self.executeStatemnt(query) { statement -> Void in
+                    let bindings: [SQLite.SQLiteValue] = [
+                        .string(identifier.databaseKey()),
+                    ]
+                    try statement.bind(bindings)
+                    try statement.step()
+                }
+                callback(.success(()))
+            } catch {
+                callback(.failure(error))
+            }
+        }
+    }
+
+    func get(identifier: PackageCollectionsModel.PackageCollectionIdentifier,
+             callback: @escaping (Result<PackageCollectionsModel.PackageCollection, Error>) -> Void) {
+        self.queue.async {
+            do {
+                let query = "SELECT value FROM PACKAGES_COLLECTIONS WHERE key == ? LIMIT 1;"
+                let collection = try self.executeStatemnt(query) { statement -> PackageCollectionsModel.PackageCollection in
+                    try statement.bind([.string(identifier.databaseKey())])
+
+                    let row = try statement.step()
+                    guard let data = row?.blob(at: 0) else {
+                        throw NotFoundError("\(identifier)")
+                    }
+
+                    let collection = try self.jsonDecoder.decode(PackageCollectionsModel.PackageCollection.self, from: data)
+                    return collection
+                }
+                callback(.success(collection))
+            } catch {
+                callback(.failure(error))
+            }
+        }
+    }
+
+    func list(identifiers: [PackageCollectionsModel.PackageCollectionIdentifier]? = nil,
+              callback: @escaping (Result<[PackageCollectionsModel.PackageCollection], Error>) -> Void) {
+        self.queue.async {
+            do {
+                let query: String
+                if let identifiers = identifiers {
+                    // FIXME: validate that the in clause is not getting too large
+                    query = "SELECT value FROM PACKAGES_COLLECTIONS WHERE key in (\(identifiers.map { _ in "?" }.joined(separator: ",")));"
```

Comment:
Do the identifiers need to be escaped or quoted in some way, or are they guaranteed to be well-formed at this point?

Reply:
Actually, maybe I'm reading that wrong:  does each identifier get replaced with a single question mark?

Reply:
> does each identifier get replaced with a single question mark?
 
yes